### ISSUE TITLE
ipfs and pacman 

### ISSUE THREAD

**Issue by jbenet** [ipfs/notes#84]
notes integrating with pacman here

**Comment by robcat** [comment]
Relevant thread on Arch Linux forums: https://bbs.archlinux.org/viewtopic.php?id=203853

**Comment by jbenet** [comment]
@robcat thanks! let us know how we can help. we'd love to contribute to making this easy + nice. cc @whyrusleeping

**Comment by robcat** [comment]
(adding @anatol, Arch Linux developer that has shown interest [on the forums](https://bbs.archlinux.org/viewtopic.php?pid=1574186#p1574186))

So, in general ipfs can help in many ways the package distribution (distributed storage, versioning, package signing).

The low hanging fruit is the distributed storage: given a "package entry" in the pacman database, get directly the package from ipfs by hash.
A custom downloader can be already plugged in pacman using the XferCommand configuration variable (some [examples](https://wiki.archlinux.org/index.php/Pacman/Tips_and_tricks#Download_speeds))

The big problem is the following: the pacman database includes only SHA256 and MD5 hashes, and no ipfs-style multihashes (that unfortunately cannot be constructed using the already included hashes).

My plan is to build a custom XferCommand script that, querying some kind of serice, translates the SHA256 hash into a "standard" ipfs hash, and then does `ipfs add -o <cache-path> <ipfs-hash>`

This hash translation service can be centralized in this initial stage, but it adds a constant lag to each package download (the next step should be to build and distribute a package database with ipfs-multihashes).

I'm still in search of a more elegant solution that doesn't require a central translation service, ideas are welcome

**Comment by fazo96** [comment]
A possible solution could be creating a dag with all packages: For example having the repo mirror servers add everything to ipfs and provide the hash of the folder, then deduplication takes care of efficiency and the pacman client only needs that one hash of the latest version of all the packages and the .db repository files :+1:

For this to work smoothly we would probably need to patch pacman. Or maybe a tool that wraps pacman but uses IPFS to get the data could be built.

**Comment by robcat** [comment]
@fazo96 About deduplication: I downloaded the pool of archlinux packages and tried both the chunking algorithms provided by ipfs (fixed blocks and rabin). Apparently there is no detectable deduplication in my case in either case. Can you suggest an effective chunking strategy for xz compressed packages?

About the dag of packages: cool idea, but it requires to manage a fat central mirror that regularly rsyncs and compiles a new dag at every update. Unfortunately I don't have such a server available (but maybe an existing Arch mirror could be interested in running the ipfs daemon?).

**Comment by fazo96** [comment]
@robcat having an existing mirror run an ipfs daemon would be ideal: by mounting its IPNS publication using fuse, it could store the data directly in IPFS while also being able to read it from the file system and when it updates the files it would automagically update what's on IPFS.

About deduplication, there's probably no trivial way to apply it inside packages but at least two copies of the same package will produce the same hash and thus will become the same file in IPFS, so that the pacman client can get a package from any IPFS node that has a copy without having any configuration telling it the nodes' ip addresses.

This way choosing the best mirror, ranking them etc will not be needed: ipfs will be in charge of downloading the package from the best location. This will mean that updating even when an Internet connection is not there could be implemented if there's a reachable computer that is serving the packages and .db files.

**Comment by jbenet** [comment]
> @fazo96 About deduplication: I downloaded the pool of archlinux packages and tried both the chunking algorithms provided by ipfs (fixed blocks and rabin). Apparently there is no detectable deduplication in my case in either case. Can you suggest an effective chunking strategy for xz compressed packages?

Are they tarballs? Did you try `ipfs tar`? it imports tars smartly. 

@whyrusleeping what is the tar branch that detects tarballs on add?

**Comment by robcat** [comment]
@jbenet 
> Did you try ipfs tar? it imports tars smartly.

Cool feature! I didn't know about that. (at the moment it doesn't seem to work for some tars, I'm opening a separate issue)

In the Arch Linux case it's unfortunately not very useful, since the packages are signed only after compression. Distributing the packages in uncompressed form would mean:
- having to recompress the package at installation time to verify the signature at a great CPU cost
- guessing which compression settings were used by the developers on their machine, in order to reproduce the same exact tar.xz file

**Comment by fazo96** [comment]
Well, packages aren't that heavy. I mean, the biggest package I even downloaded is probably netbeans which is less than 300 MB, and it's wayyyyyy bigger than the average package.

I don't think there are many packages that share common data except multiple versions of the same package. It would be very nice to figure out how to take advantage of deduplication with the current way pacman packages stuff, but even without that, an IPFS transport for pacman would be a huge step forward in terms of efficiency and repository management.

Just having the official repository publish the hash of the latest version of the packages folder and all the mirrors just automatically pinning that when it updates would propagate changes pretty fast and no one would have to worry about choosing the fastest mirror except bitswap developers :+1: 

To sum it up my humble suggestion is not to try too hard to fix this problem (applying deduplication to arch packages) now. IPFS is still very useful in this use case, even without that, and in the future pacman could consider packaging stuff differently if IPFS becomes the standard for distribution.

**Comment by robcat** [comment]
> Just having the official repository publish the hash of the latest version of the packages folder and all the mirrors just automatically pinning that when it updates would propagate changes pretty fast and no one would have to worry about choosing the fastest mirror except bitswap developers

@fazo96 
I hear you, this would be a pretty good solution.

But the problem is, to follow your plan would mean to convince the central *official mirror* to:
- hash the whole repository using ipfs
- manage and run an ipfs node (at least to initially seed the new packages)
- use twice the storage space (to store the ipfs blocks of the whole repository)
- sign and publish the ipns periodically (right now there's no central authority, only individual packages are signed)

There is a problem of incentives here: why should the central authority to do all this, if nobody is already using ipfs?

**Comment by pbronez** [comment]
> convince the central official mirror 

@robcat shouldn't any mirror work, as long as IPFS users trust it? Anyone could set up a mirror to serve as the bridge to IPFS. Then once people start using it, down the line the official repo could take over that function.

Setting up a new mirror to copy packages into IPFS is closely related to the archival efforts discussed at https://github.com/ipfs/archives/issues/5

**Comment by fazo96** [comment]
> manage and run an ipfs node (at least to initially seed the new packages)

It's not hard at all to do, there are also docker images.

> use twice the storage space (to store the ipfs blocks of the whole repository)

You can mount IPNS with FUSE so that you don't need twice the space

> sign and publish the ipns periodically (right now there's no central authority, only individual packages are signed)

Well, if the packages are all in one folder, and that folder is inside the IPNS mountpoint (using FUSE), when a file changes everything gets republished

> @robcat shouldn't any mirror work, as long as IPFS users trust it? Anyone could set up a mirror to serve as the bridge to IPFS. Then once people start using it, down the line the official repo could take over that function.

Yes :+1: I don't see why a random guy can't set up an IPFS mirror. Also no need to trust IPFS or the mirror since the packages are signed.

**Comment by robcat** [comment]
> I could do it, but I only have 1 Mbit upload bandwidth, for four people and at least 6 devices...

Ok, so what about this three pronged approach:
- @fazo96 publishes the ipns entry (minimal bandwidth required)
- I keep a non-authoritative mirror with a lot of bandwidth that will sync the most requested packages
- users will put https://ipfs.io/ipns/<faso96-hash>/ at the top of their mirrorlist

My node is already up at 178.62.202.191 and it independently adds all the x86_64 packages from core, extra, multilib, testing and multilib-testing (it syncs to an official arch mirror and adds the packages to ipfs).

@fazo96 you can go ahead and publish the ipns entry, if IPFS works the bulk of the requests will not even hit your node :)

**Comment by fazo96** [comment]
@robcat I'm very interested in trying this out! :) 

However:

- I need the hash to be able to publish something
- You can publish it yourself if you want, or you can mount ipns with FUSE and drop the files in `<mountpoint>/local` so that they are automatically published on your node's IPNS (this is the suggested approach) while also being accessible on the filesystem thus not taking up twice the space!
- users shouldn't use ipfs.io as the gateway or all the traffic will have to come from ipfs.io to the users and it doesn't make much sense: they would have to use their local gateway (localhost)
- I'm not sure what the file structure of a pacman mirror is, so I need to read about that so that we can try this out and figure out the best way to publish everything
- I have quite some credit on digital ocean thanks to github's student developer pack, and I'm willing to set up a mirror for pacman using IPFS :+1: 

Also, another use case for this is local package sharing: I have two computers in LAN and a shitty bandwidth to the internet, and all available solutions to share packages in LAN are pretty ugly compared to this.

If you want we can meet on freenode (I'm `Fazo`) later today or some other time to try this out :+1: If we succeed (we probably will) we can set up a page on the arch wiki or a github repo that explains how to use IPFS with pacman this way (without any external tools except ipfs)

**Comment by jbenet** [comment]
If you have troubles with IPNS bug me directly. we'll improve the perf to fix it for you.

**Comment by fazo96** [comment]
@jbenet thanks a lot, but I think with the 0.3.10 caching we'll be fine :+1: 

It would be cool if records would be kept for a long time if nothing more recent is available, not sure if that's already in the codebase.

**Comment by jbenet** [comment]
@fazo96 right now records expire. i do hear you though, most IPNS users today would prefer non-expiring records. my crypto paranoia could be deferred.

**Comment by anatol** [comment]
> Also, another use case for this is local package sharing: I have two computers in LAN and a shitty bandwidth to the internet

@fazo96 A bit of out-of-topic. I was looking for a simple Arch package sharing tool for LAN and did not find what matches my expectations. So I wrote my own tool - https://github.com/anatol/pacoloco an Arch package proxy repo. It is not published anywhere and just a bunch of code without much documentation. But it works well for several months for me. Advantages of this tool - pure C, no external dependencies except glibc, run perfectly at OpenWRT MIPS router, single-threaded, event-loop based architecture, HTTP pipeline support to reduce request latency.

**Comment by fazo96** [comment]
@anatol Cool, if you could set up some documentation on how to use it and maybe even a PKGBUILD to the AUR, I'd use it.

**Comment by fazo96** [comment]
Ok, time for some results!

me and @robcat did it. He set up an arch linux mirror on a VPS then we got it to store files inside the FUSE mountpoint (`/ipns/local`) so that they are always available via IPNS.

### Using IPFS to download arch packages

The node that is serving them is `QmPY3tsmwoCjePd9SEudAGX2bZU65SwiL8KBKaSRKyUdzt`. If you want to try it, you can put `Server = http://localhost:8080/ipns/QmPY3tsmwoCjePd9SEudAGX2bZU65SwiL8KBKaSRKyUdzt/$repo/os/$arch` on top of your pacman mirrorlist. Make sure your daemon is running, then it will download packages from IPFS when possible! _It's that simple_.

It has a few hiccups, for example the timeout is too low and if it's your first time downloading package it's going to fall back to boring centralized mirrors. Also the mirror we have is incomplete due to the resources needed to hash tens of gigabytes of files.

__But it works!__ Keep in mind that the IPFS mirror mentioned was just used for testing, we are not guaranteeing anything.

### Setting up a mirror

Just follow the steps to set up a regular arch mirror, but mount IPNS to `/ipns` before and then drop your mirror files in `/ipns/local/`. That's it, _it's that simple_. go-ipfs + FUSE works fine on ubuntu 15.10 and Arch Linux.

**Comment by jbenet** [comment]
This is awesome! congrats @robcat and @fazo96 ! \o/

![](http://rs773.pbsrc.com/albums/yy19/spockachu/gifs/clap.gif~c200)

---

Please take a look at @diasdavid's recent `registry-mirror` work. It has very similar goal (replicate npm), and has been addressing many of the issues. 

- You can see a _preview_ blog post here: https://ipfs.io/ipfs/QmREB6yt38NWQVjjaeLKYdhotLqmvt421GxBT2J1UDB8uU/blog/4-registry-mirror/ (not yet published to our blog)
- https://github.com/diasdavid/registry-mirror (includes how it works)
- talk went live yesterday here: https://www.youtube.com/watch?v=-S-Tc7Gl8FM
- some of the development discussion here: https://github.com/ipfs/notes/issues/2

@diasdavid uses the new `mfs` (files) API in `0.4.0-dev`, to make life much easier. I suspect you can likely use that too programmatically, instead of relying on the `/ipns` mount. then can publish the hash when you're done (`ipfs name publish $hash`). ipfs's fuse interface is not the best of things yet (apologies there, if you found UX issues), so definitely give the files api a try.

Also `0.4.0` is way faster than `0.3.x`.

**Comment by fazo96** [comment]
@jbenet Thanks!!

We know about `0.4.0` being faster and the issues with the mount, but we wanted to make it as easy as possible for arch mirror maintainers to integrate IPFS in the mix, and the IPNS mounts really helps because it makes it very simple to use the regular arch mirror administration tools and practices with IPFS. However, we'll surely consider a better approach and test some more.

**Comment by jbenet** [comment]
> Just follow the steps to set up a regular arch mirror, but mount IPNS to /ipns before and then drop your mirror files in /ipns/local/. That's it, it's that simple. go-ipfs + FUSE works fine on ubuntu 15.10 and Arch Linux.

I have to say, it is **absolutely awesome** that it's this simple. :heart: to all the hard work making the abstractions nice. (cc @whyrusleeping)

@robcat @fazo96 -- @whyrusleeping and i have discussed making the `files` api mountable as well. that interface still needs to be defined. potentially something like this https://github.com/ipfs/go-ipfs/issues/2060#issue-121849761

**Comment by jbenet** [comment]
> We know about 0.4.0 being faster and the issues with the mount, but we wanted to make it as easy as possible for arch mirror maintainers to integrate IPFS in the mix, and the IPNS mounts really helps because it makes it very simple to use the regular arch mirror administration tools and practices with IPFS. However, we'll surely consider a better approach and test some more.

That's fantastic. it really is :heart:

@fazo96 we may want to make a simple `sh` script or `Makefile` that does all of this

- boilerplate shell script params stuff: 
  - https://github.com/jbenet/ipfs-screencap/blob/master/ipfs-screencap (`bash`)
  - https://github.com/jbenet/ipfs-paste/blob/master/ipfs-paste (`sh`)
- install ipfs with:
  - simple shell script https://github.com/ipfs/install-ipfs
  - new hotness https://github.com/ipfs/ipfs-update
- boilerplate Makefile with install things: https://github.com/ipfs/go-ipfs/blob/master/test/Makefile

**Comment by fazo96** [comment]
We're still figuring out issues with rsync crashing go-ipfs when used with a target inside `/ipns/local`. If we figure out exactly what is causing the problem, we'll open an issue about it :+1:

**Comment by robcat** [comment]
Ok, I've been able to sync the "core" repository to /ipns/local using this script, and it works wonderfully on my Arch machine with ipfs 0.3.10:
https://gist.github.com/robcat/3dbbafee096269b6843a

But me and @fazo96 were trying to make it work on a Ubuntu 15.10 VM (same versions of ipfs and fuse), and rsync doesn't detect the unchanged files (it overwrites every file, triggering the re-hash and re-publish by ipfs). This is of course too slow.
Unfortunately this problem is not straightforward to debug (it seems to be a timeout issue in the interaction between rsync and fuse): if anyone want to try, please contact me or @fazo96 in order to get access to the VM.

@jbenet The mountable files api you described can be useful in a lot of situations.
But in the "mirror" use case, we are relying on the auto-publishing feature of /ipns/local. I believe we would have no real advantage in copying the packages in a different folder.

**Comment by whyrusleeping** [comment]
@robcat i think I know the issue. the Atime and Ctime of the files in /ipns arent set. Let me work on a patch for ya

**Comment by whyrusleeping** [comment]
hrmm... after a bit more thought its not going to be easy to fix. I'm working modtimes into the mfs code on 0.4.0 (the code behind the fuse interface). You could try the `-c` option on rsync as a workaround for now

**Comment by robcat** [comment]
> hrmm... after a bit more thought its not going to be easy to fix. I'm working modtimes into the mfs code on 0.4.0 (the code behind the fuse interface). You could try the -c option on rsync as a workaround for now

Uhm, I double checked and in fact there is something seriously wrong with my Ubuntu machine.
Rsync doesn't give any errors, but the resulting files accessed via ls have zero sizes (of course the -c option doesn't help):

```
# ls -l /ipns/local/foo
total 0
-rw-rw-rw- 0 root root 0 Aug 30  1754 acl-2.2.52-2-i686.pkg.tar.xz
-rw-rw-rw- 0 root root 0 Aug 30  1754 acl-2.2.52-2-i686.pkg.tar.xz.sig
[...]
```

But the file objects are stored correctly by ipfs:
```
# ipfs/ipfs object links QmcCShzUmGvMjbWyC1jT1fG9y4LFYt7Cxwms7Pm82vg72d
Hash                                           Size   Name                                            
QmaKpUC352G5raBe1YMhKAWEGR4zbHVh4GgWG4sbVuEdFo 133358 acl-2.2.52-2-i686.pkg.tar.xz                    
QmaobnVG1xfHcYEF4DiLQYQFGGyJaZ32rYNUCeAchGqXe5 607    acl-2.2.52-2-i686.pkg.tar.xz.sig    
```

To ask for a third opinion, I will try soon with osx.

**Comment by fazo96** [comment]
@robcat that's very nice!

I'll give it a try too on my machine if I have the time and network bandwidth. We could try setting up a dockerfile based on the go-ipfs one that also mounts fuse and keeps an updated arch mirror using your script, so that we can run it on any platform.

Then, only the "root" arch mirror using IPFS needs to actually do this. The other mirrors will just need to pin the IPNS name of the "root" mirror once pub/sub is implemented. As a temporary workaround, a very small script could poll for IPNS updates and pin the new hash.

**Comment by whyrusleeping** [comment]
@robcat you could try enabling the fuse debug logs, it currently need to be added in and recompiled like so:
```diff
diff --git a/fuse/ipns/ipns_unix.go b/fuse/ipns/ipns_unix.go
index bd4b861..5f634d4 100644
--- a/fuse/ipns/ipns_unix.go
+++ b/fuse/ipns/ipns_unix.go
@@ -23,6 +23,12 @@ import (
        ft "github.com/ipfs/go-ipfs/unixfs"
 )
 
+func init() {
+       fuse.Debug = func(i interface{}) {
+               fmt.Println(i)
+       }
+}
+
 var log = logging.Logger("fuse/ipns")
 
 // FileSystem is the readwrite IPNS Fuse Filesystem.
```

but that should give you some more insight

**Comment by atondwal** [comment]
So I tried @robcat's script on my machine and it works for `[core]`, but the bigger packages from `[community]` (in my case `0ad`, it's really big, but it's also alphabetically first) run into ipfs/go-ipfs#1456 and crashes violently, complaining about too many open files.

When it does work, though, it works beautifully. I'm playing with it at `/ipns/ani.sh/`, but that's my laptop, so don't expect much uptime.

**Comment by whyrusleeping** [comment]
@atondwal thanks for the report! could you try upping your ulimit before starting the daemon and seeing how things go for you then?

On linux thats: `ulimit -n 2048` (or any number you like,really)

**Comment by atondwal** [comment]
Huh, that makes it fail to mount
```
~ ❯❯❯ ipfs daemon --mount                                                                                                                                                                                                                  ⏎
Initializing daemon...
Swarm listening on /ip4/127.0.0.1/tcp/4001
Swarm listening on /ip4/192.168.0.9/tcp/4001
Swarm listening on /ip4/68.100.248.158/tcp/20207
Swarm listening on /ip6/::1/tcp/4001
API server listening on /ip4/127.0.0.1/tcp/5001
Gateway (readonly) server listening on /ip4/127.0.0.1/tcp/8080
01:53:18.061 ERROR    bitswap: failed to find any peer in table workers.go:92
01:53:18.144 ERROR core/comma: error mounting: %!s(<nil>) failed to find any peer in table mount_unix.go:219
Error: failed to find any peer in table
```
(without `--mount` it starts fine though...)

**Comment by atondwal** [comment]
Oh, I just had to `umount` it by hand before remounting it, because I had a shell open at that directory when I restarted `ipfs daemon`.

Also after playing around with the, the problem wasn't t he filesize, it was the fact that I was using `cp` instead of `rsync`. Which is sort of a strange thing to be the problem.

**Comment by atondwal** [comment]
Ah, the problem was with FUSE. Just doing `ipfs add -r` works like a charm.

**Comment by Kubuxu** [comment]
@atondwal yes FUSE has its quirks, it is better to use CLI for the most part.

**Comment by Split7fire** [comment]
Have read an entire thread, but dont understand:

- ipns is not an vialable solution?
- someone say to use mfs instead, where can I read about it?
- any chance for _official_ support for Archlinux?

**Comment by Kubuxu** [comment]
@Split7fire I think IPNS is viable solution for releasing updates to the repo. IPNS via FUSE might not be the best solution as it isn't at its best performance yet.

`mfs` (mutable file system) was working name for Files API of IPFS. It is available with `ipfs files` command and allows for creation, modification and deletion of files and directories under IPFS (everything is still immutable but changes in hash are pushed up the DAG).

Chances for official support right now are small, but using repo via http bridge or custom pacman is quite possible.

**Comment by fazo96** [comment]
@Kubuxu yeah I can confirm that works quite well, the problem is that we don't have anyone with the time and resources to maintain such a mirror.

**Comment by Split7fire** [comment]
Any (less size) alternative with ipfs Files API?

**Comment by Split7fire** [comment]
Accidentally found [ipget utility]( https://github.com/ipfs/ipget) which "probably" can help. Some posts ago people said that using utility such wget in pacman.conf would be a reasonably well solution. Any thoughts?

**Comment by fazo96** [comment]
@Split7fire could work, but you need someone hosting a full (or at least partial, for testing) arch repository mirror on IPNS. Then, you could add it as a mirror in the pacman configuration and tell it to use ipget for downloading. If ipget is correctly called by pacman and it provides the files fast enough not to trigger any timeouts, then it'll work 👍

**Comment by Split7fire** [comment]
For testing purposes I can host arch repo for a while :) @fazo96 Any howto will be appreciated.

**Comment by fazo96** [comment]
@Split7fire off the top of my head, I remember [this section](https://wiki.archlinux.org/index.php/Pacman/Tips_and_tricks#Network_shared_pacman_cache) of the Arch Wiki, it tells you how to share a local package cache. Starting from there, if you also share the repositories (.db files) and you have every single package in the cache for both architectures then you're hosting a full mirror. A full mirror takes around 50 GB of space and needs to be synced often.

You can find more details about how to host a mirror [here](https://wiki.archlinux.org/index.php/DeveloperWiki:NewMirrors).

The point is, if someone were to keep all the mirror data on IPFS, every IPFS HTTP gateway would turn into a full arch linux repository mirror. The problem is that you also need a copy of the whole thing on a regular file system so that you can keep it updated with rsync, so you either mount IPNS with fuse and hope it's stable enough, or you need two copies of the whole mirror upping the space required to around 100 GB I guess.

If you manage to get a full HTTP mirror working (or even partial), I can help you set up the required scripts to keep it on IPFS.

**Comment by Kubuxu** [comment]
I would love to help, as I have arch mirror in my local network already (it is 61GiB), but unfortunately go-ipfs is not able to punch the NAT I am behind.

**Comment by chpio** [comment]
ok, so i did an rsync copy from an TIER2-Arch-mirror and added the dir into ipfs (with the default `ipfs add -r .` command), but im getting errors now from pacman and also when i try to open a file directly:

> internalWebError: cannot currently read symlinks

https://ipfs.io/ipfs/QmdhPp1TSS2pLfNbAsTMPbnHtqA8i9oo5CypwhvKkar6Qp/core/os/x86_64/acl-2.2.52-2-x86_64.pkg.tar.xz.sig

What's the best way to add all the files and dirs as they are added right now, but resolve the symlinks, so they are added just like normal files?

ps: 100GB shouldn't be a real problem in the modern days where you can get a 5TB HDD for a few $/€/$whatever

**Comment by Kubuxu** [comment]
Gateway is not supporting symlinks, which will be a problem in this case, there are three solutions for that, add symlink

**Comment by Split7fire** [comment]
you can replace symlinks by their referral during rsync

**Comment by chpio** [comment]
@Split7fire yeah, the disk usage went from 55GB to 126GB (just for the rsync copy). rsync is also missing an option to convert symlinks to reflinks D: (https://bugzilla.samba.org/show_bug.cgi?id=10170) the `ipfs add` time went up also. a symlink-aware adder script (which would cache symlinks, so they wouldn't be added twice to ipfs, resulting in faster adds...) would be much nicer

**Comment by fazo96** [comment]
Wow @chpio, that's great! I'll try it as soon as I can!

**Comment by chpio** [comment]
nice, but take in mind that this was just a test and im not updating it (it's already outdated)

**Comment by photonxp** [comment]
An ipfs repo of OS mirror could possibly use too much disk, I/O, bandwidth, etc. on a personal computer. But a repo of a specific programming language use considerably less resources, so it's possible to keep a repo of the language even on low-end laptops. So don't get intimidated, just start with something small.

**Comment by chpio** [comment]
On 06.02.2017 05:27, photonxp wrote:
> An ipfs repo of OS mirror could possibly use too much disk, I/O, bandwidth, etc. on a personal computer. But a repo of a specific programming language use considerably less resources, so it's possible to keep a repo of the language even on low-end laptops. So don't get intimidated, just start with something small.  
> 

uhh yeah, im running this now:
```
# /etc/pacman.d/mirrorlist
Server =
http://127.0.0.1:8081/ipns/QmaeXrsLHWm4gbjyEUJ4NtPsF3d36mXVzY5eTBQHLdMQ19/$repo/os/$arch
```

and this to fix the timeouts:
```
# /etc/pacman.conf
XferCommand = /usr/bin/wget --passive-ftp -c -q --show-progress -O %o %u
```

**Comment by atommixz** [comment]
@chpio works for me only with port 8080, clean installation ipfs-0.4.5 and ipfs init, I mean:
# /etc/pacman.d/mirrorlist
Server = http://127.0.0.1:8080/ipns/QmaeXrsLHWm4gbjyEUJ4NtPsF3d36mXVzY5eTBQHLdMQ19/$repo/os/$arch

strange behavior for wget, idk real speed --.-KB/s

**Comment by chpio** [comment]
> works for me only with port 8080. I have clean installation ipfs-0.4.5 and ipfs init, I mean

yeah you have to change it to the port you configured for the ipfs-http-gateway

**Comment by wanderer** [comment]
@chpio do you intend to keep `/ipns/QmaeXrsLHWm4gbjyEUJ4NtPsF3d36mXVzY5eTBQHLdMQ19/` updated for the foreseeable future?

**Comment by jruere** [comment]
AFAIU, https://github.com/ipfs-filestore/go-ipfs/tree/master/filestore could help a lot avoiding duplicating storage requirements.

PS: I could not find a peer with an up-to-date repository.

**Comment by whyrusleeping** [comment]
@jruere that functionality exists in ipfs now, see: https://github.com/ipfs/go-ipfs/blob/master/docs/experimental-features.md#ipfs-filestore

**Comment by jruere** [comment]
@whyrusleeping that's very neat! Sorry I missed it.

Do I understand correctly that this would simplify adding IPFS to an existing mirror?

**Comment by whyrusleeping** [comment]
@jruere Yeah, it should make things a bit simpler. The caveat is around updating things, cc @kevina for discussion around updating files that are in the filestore.

**Comment by jcaesar** [comment]
I've experimented around a bit with @chpio's repo and found it to be unbearably slow. I set up a cronjob that resolves ``/ipns/QmaeXrsLHWm4gbjyEUJ4NtPsF3d36mXVzY5eTBQHLdMQ19/`` hourly and pushes the result to ``_dnslink.arch-chpio.cameo.liftm.de``. Using ``http://arch-chpio.cameo.liftm.de:8080/$repo/os/$arch`` (depends on ipfs running on ``localhost:8080``) is a bit faster and practically bearable, imho.

PS: (I know it's been years, but I feel like this should be mentioned)
@robcat:
> having to recompress the package at installation time to verify the signature (at a great CPU cost)

From what I hear, recompressing files and getting the exact same binary is fragile at best. This may not be an option.

**Comment by robcat** [comment]
@jcaesar 
> From what I hear, recompressing files and getting the exact same binary is fragile at best. This may not be an option.

There is an ongoing effort in [deterministic builds on Arch Linux](https://tests.reproducible-builds.org/archlinux/archlinux.html) (77% done), and the compressing part of the project is the easiest one (compiling deterministically is much harder).

**Comment by chpio** [comment]
> @chpio do you intend to keep /ipns/QmaeXrsLHWm4gbjyEUJ4NtPsF3d36mXVzY5eTBQHLdMQ19/ updated for the foreseeable future?

@wanderer Nope, im going to shut it down in a few days.

**Comment by tobixen** [comment]
So there is no resolution on this?

I have several laptops and other computers running arch linux, and a thin uplink - it's a shame to download the same packages over and over again.  Also, some non-standard packages aren't available at the standard mirrors and may be really slow to download even when having a decent uplink, IPFS is just perfect for distributing such content.

I tried putting the /var/cache/pacman/pkg download directory under NFS for a while before learning about IPFS, but it was painful - remembering to mount it up when the laptop is at home and to unmount before leaving home, etc, not giving any benefits i.e. when being at vacation having to upgrade two laptops over mobile networks, etc.  IPFS seems to be the perfect solution.

I think the "quick and dirty" solution would be to just ensure /var/cache/pacman/pkg being a ipfs mount - though, I'm having problems mounting up ipfs currently, so I haven't gone that route yet.  I believe the proper solution is native support from arch linux, making sure ipfs hashes for the uncompressed tar files are available and that pacman can download it.

**Comment by whyrusleeping** [comment]
No resolution yet, And this is something I would personally love to have too.

If anyone is interested in spending some time developing a proper solution here, let me know.

**Comment by simonbru** [comment]
@tobixen I also think using IPFS (or another P2P distribution mechanism) for linux packages is a cool solution. I hope people will continue to work on it.
However if your main use case is to share packages in your LAN, I suggest you to try [pacserve](https://wiki.archlinux.org/index.php/Pacserve) in the meantime. It does not require much configuration of maintenance and works pretty well in my experience.

**Comment by jcaesar** [comment]
@chpio: I don't know if I have the resources to run something like this, but could I ask you to share whatever scripts or configuration you have that made the mirror work?

**Comment by VictorBjelkholm** [comment]
I've made some reusable scripts for pulling down a repository and adding it to IPFS. https://github.com/VictorBjelkholm/arch-mirror

- https://github.com/VictorBjelkholm/arch-mirror/blob/master/find-fastest-mirror.sh helps you find a fast mirror to sync from
- https://github.com/VictorBjelkholm/arch-mirror/blob/master/syncrepo.sh downloads all the files from the mirror (here with the rsync command we can maybe find a parameter that can make rsync copy files instead of copy the symlink?)
- https://github.com/VictorBjelkholm/arch-mirror/blob/master/ipfsify.sh setups IPFS repository with bunch of configuration to make the mirror smaller/faster to add to ipfs. Also adds the final hash to IPNS and so on. DNS update will probably happen in here too.
- https://github.com/VictorBjelkholm/arch-mirror/blob/master/run-daemon.sh starts a ipfs daemon local to the directory. Has pubsub to make IPNS faster.


Still, IPFS doesn't handle symlinks and this cannot be used for pacman just yet. But now we have some scripts to work from.

**Comment by Kubuxu** [comment]
@VictorBjelkholm You can try initializing the repo with `badgerds` profile (`ipfs init --porfile=badgerds`) it should be even faster then.

**Comment by VictorBjelkholm** [comment]
Thanks @Kubuxu, I'll try that in a future update.

## Update

I've successfully managed to download, add to IPFS and use it for a full system upgrade already :tada: 

The full instructions for usage and hosting your own mirror is over here: https://github.com/VictorBjelkholm/arch-mirror

**TLDR of usage is**: Add `Server = https://ipfs.io/ipns/arch.victor.earth/$repo/os/$arch` in the top of `/etc/pacman.d/mirrorlist`

This DNS record gets updated at least once per day, I have the sync setup to happen automatic twice a day. It's currently hosted on at least three different machines, my home server, a digitalocean droplet specifically for this arch-mirror and also `pollux` which is part of our standard infrastructure. I'll also pin this on ipfs-cluster once we've done a bit more testing.

If you want to help out now, please try the server above and report any issues either here on on https://github.com/VictorBjelkholm/arch-mirror/issues/new

If people are stepping up and want to help out maintaining the scripts, we can move the repository over to `ipfs-shipyard` instead of within my own Github scope.

**Comment by VictorBjelkholm** [comment]
> I set up http://127.0.0.1:8080/ipns/QmaeXrsLHWm4gbjyEUJ4NtPsF3d36mXVzY5eTBQHLdMQ19/$repo/os/$arch at the top of the mirror list - this option seems not to be covered in your README? Seems to work, except frequent fallbacks to the other mirrors due to timeouts - I should probably see if I can adjust up the timeout.

Yeah, it's mentioned in the readme here: https://github.com/VictorBjelkholm/arch-mirror/blob/master/README.md#usage-with-local-daemon

Let me know if you figure out how to change the timeout, as it would be good to include in the readme.

> Wouldn't /ipns/arch.victor.earth/$repo/os/$arch work out when ipfs-mount is used? (I haven't had much luck with ipfs mount - Error: fuse failed to access mountpoint /ipfs despite ownership and permissions apparently being right - but that's a digression)

Hm, I haven't had much luck with fuse mounts lately either, which is why I left that out. Fuse support in go-ipfs is notably in need of some love: 

> Will one gain any benefits at all using https://ipfs.io/ as compared to a conventional mirror?

Yeah, better caching than traditional mirrors (assumption, haven't actually looked into how other Tier 2 mirrors does caching yet) and verifiable packages. It's also part of our migration process, so we can easily move to something else in the future. For example, ipfs.io could resolve to localhost once every computer runs IPFS itself. But best is of course to already run your own go-ipfs instance locally.

> IPFS hashes is immutable as far as I've understood (and that's the reason why we need ipns?) - so with the suggestion to use /ipfs/QmQgfsgwenh3tiCmgbhtxr9sAVCaxbBRFyYu1GjAp84rgo you'll get a permanently frozen "mirror"? ("harder for you to update" seems like an understatement?)

Yes, they are immutable and why we need IPNS or DNS. The suggestion means (but maybe not clear enough) that you would check either the readme of arch-mirror (I update that hash everytime I run the update) or resolve the IPNS/DNS record and replace it manually in your `mirrorlist`. If you keep the same hash forever, it'll be permanently frozen, that is correct. And you would have to manually update it to point to the new hash once you see it.

**Comment by tobixen** [comment]
I spent some time searching, but did not find any obvious way to increase the timeout.

**Comment by jruere** [comment]
Is this not enough?

`XferCommand = /usr/bin/wget --timeout=180 -c -O %o %u`

**Comment by tobixen** [comment]
Usage of the XferCommand breaks the "TotalDownload"-feature in pacman, I had hoped there were some simple/obvious way to increase the timeout without resorting to that one.

Another workaround: It seems like the same URL can be specified several times in the pacman mirror list ...

    $ grep 127.0.0.1 /etc/pacman.d/mirrorlist
    Server = http://127.0.0.1:8080/ipns/arch.victor.earth/$repo/os/$arch
    Server = http://127.0.0.1:8080/ipns/arch.victor.earth/$repo/os/$arch
    Server = http://127.0.0.1:8080/ipns/arch.victor.earth/$repo/os/$arch

This doesn't seem to help me though, if it times out at the first one, it also times out at the two others.

Maybe it's just good to skip IPFS and go to some other mirror when IPFS times out.  To help the IPFS project without going full-scale with running a full mirror, I've thrown up this cronjob as root:

    33 * * * * sudo -u ipfs ipfs add --pin=false /var/cache/pacman/pkg/* ; find /var/cache/pacman/pkg -mmin +30 -delete

I wish ipfs mount had been working ... having /var/cache/pacman/pkg working directly from ipfs would be super.

**Comment by tobixen** [comment]
@VictorBjelkholm is your mirror still functioning?  I get timeouts all the time now.  Could I have the "swarm address" to the mirror so I can add it to my peer lists?

**Comment by VictorBjelkholm** [comment]
@tobixen hm, that's weird. It's currently hosted at two nodes, one I have running at home and one at Digital Ocean. The addresses:

```
/ip4/88.18.104.235/tcp/7000/ipfs/QmUWA5FD6otih6aCqnSTSy4cQbcttwCYpA6mYMPJ9qAyKi
/ip4/167.99.95.100/tcp/4001/ipfs/QmX1WWwNwfEsuzsqWFVstnEkzVFN5i4kefqJQMb5S4umAT
```

Probably need to setup some integration with our ipfs cluster we have running, and pin it there on every update. I've checked the logs and have been running fine since the inception.

**Comment by jruere** [comment]
I also had this problem some time ago and have up on the approach. :(

Ipfs doesn't seem mature enough for me.

**Comment by tobixen** [comment]
For me, ipfs on ipv6 works out quite well sometimes.  ipfs on NAT and ipv4 sometimes doesn't work at all.  I just changed ISP at home, the new ISP is supposedly supporting ipv6, but in practice my ipv6 setup at home has been down since I changed provider.  I will try to do some research on this.

Another thing I found is that it's the network discovery that takes time.  Manually connecting to the "right" peers either in the bootstrap configuration or through the "ipfs swarm connect"-command works out wonders.

I wonder if it's any way to mirror all the metadata without running a full package mirror?  Due to the cronjob I set up above, I often have all the packages I need available locally on ipfs nodes in my house, but since I don't know the hash for them (due to ipfs timeouts) it's still needed  to download all the packages from some official mirror.

It seems to me that ipfs is currently not very well suited for "the long tail" of content, when being the only node searching for some content it may be impossible to find it.  If many nodes are searching for the same content, ipfs probably works great.  Meaning that the more people that uses the mirror, the better the performance will be.

**Comment by tobixen** [comment]
For the reference, I did this:

    ipfs swarm connect /ip4/88.18.104.235/tcp/7000/ipfs/QmUWA5FD6otih6aCqnSTSy4cQbcttwCYpA6mYMPJ9qAyKi 
    ipfs swarm connect /ip4/167.99.95.100/tcp/4001/ipfs/QmX1WWwNwfEsuzsqWFVstnEkzVFN5i4kefqJQMb5S4umAT

and then upgraded the system - packages were downloaded from ipfs without any problems whatsoever!

Maybe an idea to add a hint about this in the README.

**Comment by whyrusleeping** [comment]
poor content discovery in the past ~3mo is very likely due to recent degradation in DHT performance. We're working on improvements here, see [this issue](https://github.com/libp2p/go-libp2p-kad-dht/issues/88) for details.

**Comment by whoizit** [comment]
> For the reference, I did this:
> 
> ```
> ipfs swarm connect /ip4/88.18.104.235/tcp/7000/ipfs/QmUWA5FD6otih6aCqnSTSy4cQbcttwCYpA6mYMPJ9qAyKi 
> ipfs swarm connect /ip4/167.99.95.100/tcp/4001/ipfs/QmX1WWwNwfEsuzsqWFVstnEkzVFN5i4kefqJQMb5S4umAT
> ```
> 
> and then upgraded the system - packages were downloaded from ipfs without any problems whatsoever!
> 
> Maybe an idea to add a hint about this in the README.

```
$ ipfs swarm connect /ip4/88.18.104.235/tcp/7000/ipfs/QmUWA5FD6otih6aCqnSTSy4cQbcttwCYpA6mYMPJ9qAyKi 
Error: connect QmUWA5FD6otih6aCqnSTSy4cQbcttwCYpA6mYMPJ9qAyKi failure: dial attempt failed: 1 error occurred:
	* <peer.ID Qm*jJUN2u> --> <peer.ID Qm*9qAyKi> (/ip4/88.18.104.235/tcp/7000) dial attempt failed: dial tcp4 0.0.0.0:4001->88.18.104.235:7000: i/o timeout
$ ipfs swarm connect /ip4/167.99.95.100/tcp/4001/ipfs/QmX1WWwNwfEsuzsqWFVstnEkzVFN5i4kefqJQMb5S4umAT
Error: connect QmX1WWwNwfEsuzsqWFVstnEkzVFN5i4kefqJQMb5S4umAT failure: dial attempt failed: 1 error occurred:
	* <peer.ID Qm*jJUN2u> --> <peer.ID Qm*S4umAT> (/ip4/167.99.95.100/tcp/4001) dial attempt failed: dial tcp4 0.0.0.0:4001->167.99.95.100:4001: i/o timeout
```

**Comment by Stebalien** [comment]
@whoizit those peers appear to be offline.

**Comment by guysv** [comment]
I wrote a simple pacman hook that uploads newly-updated packages onto IPFS:
https://github.com/guysv/ipfs-pacman-cache-hook
This would allow people to support kickstarting the 'Arch-on-IPFS' idea while still using the old mirrors.

But, we still need somebody to maintain an updated packages index on IPFS, else the uploaded material won't be discoverable.

**Comment by RubenKelevra** [comment]
@tobixen what happened to the project I would love to use it :)

---

Another question arise from the discussion:

Wouldn't be a IPFS cluster not more suitable for this application?
Every maintainer (trusted peer) can alter the pinset, while the the mirrors spread the content reliably an guaranteed redundancy - without having to hold the whole repositories on each server.

This cuts out the central point of failure of the one person maintaining a pinset, in favour of every maintainer has the rights to push new updates to the cluster.

It also avoids that there are different "sync" states on different mirrors which run sync scripts independently.

For example: if you get your database from the second entry in the mirror list, because the first one was too slow to respond, every download request to the first server fails since the first server synced a long time ago.

**Comment by guysv** [comment]
@RubenKelevra I totally agree! Thats was my first thought after reading the about the new cooperative clusters. Someone should take up the glove.

**Comment by RubenKelevra** [comment]
@guysv the only question left is: How does regular computers access the cluster. When I understood the documentation right (I have not used it yet), than you have to participate on the cluster to have access

This works fine, if you have multiple machines which needs access to the data and they participate on the cluster. But in our case we have many thousands of machines which don't want to be part of the cluster, but access the data in it, caching it and exchanging data they downloaded anyway with other clients on a local network.

So as I understand we need a additionally class of nodes, like Edge Nodes or Client Nodes, which have access to the private cluster, but no duty to store any data and a setting which forbids them to send data to other parts of the network, which are not in a specified IP range.

The later would else result in machines with poor internet connections to offer data to other peers on the network.

If you run a server with a good connection you might want to lift the no-upload restriction, to help the cluster to cache data.

This way we might be able to exchange cached packages very locally, like searching for the peer with the nearest IPv6, to just avoid long routes or measure the response time/hops between different peers etc.

**Comment by jcaesar** [comment]
@RubenKelevra If you're on local network with the IPFS host anyway, you might save the storage costs for `/var/cache/pacman` and simply access everything via the HTTP gateway like a normal pacman mirror? That is, of course, only possible if you intend to keep a full mirror in IPFS.

**Comment by RubenKelevra** [comment]
@jcaesar Running a IPFS cluster member locally would result in a lot of traffic, since you're now part of the cluster and have to store some off the data.

This is impossible on slow internet connections.

I don't think there's currently a way to hold a full mirror of the cluster without taking part in the cluster.

---

But you could run http gateways on the cluster servers, like we currently do on the mirrors anyway - that would be also the way to get backwards compatibly with a cluster:

When some cluster members would offer http access to the cluster we could have Pacman access the cluster without any change on the systems.

But this would void a lot of advantages of IPFS, when the individual machines cannot take advantage of sharing the data already downloaded data locally.

It would also create a lot of additionally traffic on cluster members: When you access data which isn't locally stored, it has to be fetched from other members of the cluster, before it can be send via http, removing nearly all advantages of a sharding in the first place.

**Comment by jcaesar** [comment]
> I don't think there's currently a way to hold a full mirror of the cluster without taking part in the cluster.

You run a full mirror on one node (or a cluster of nodes), and access that through other IPFS nodes?

**Comment by RubenKelevra** [comment]
> > I don't think there's currently a way to hold a full mirror of the cluster without taking part in the cluster.
> 
> You run a full mirror on one node (or a cluster of nodes), and access that through other IPFS nodes?

I'm not sure. A cluster is a private share of data. You need to have a pre shared key to get access. I don't know if the standard client can access a cluster read-only.

Someone got more knowledge with that?

**Comment by RubenKelevra** [comment]
> > I don't think there's currently a way to hold a full mirror of the cluster without taking part in the cluster.
> 
> You run a full mirror on one node (or a cluster of nodes), and access that through other IPFS nodes?

Alright, scrap my last reply.

We could either build a private IPFS network and build a cluster on top of it, or use the public IPFS network and build a cluster on top of it.

Regardless how we decide you can access the data of the cluster via IPFS without beeing part of the cluster. In case of the private IPFS you need an key to access the network.

The cluster key is more like an identifier, who is part of the same cluster, nothing more.

So if we set up a colaborative cluster on the public IPFS, everyone can access the files and cache them with their IPFS client. The cluster on top would only make sure that the data is updated on the cluster members, when there are updates published and make sure that there are enough copies of the data within the cluster.

To access the updates (without changes to pacman) you could either add local http servers to the mirrors which accesses the IPFS via the local daemon ... or use public gateways.

You could easily setup a local cache for a network as well, without the need to have IPFS running on every machine: Just setup a IPFS node with a webserver to access the network and use this for pacman as source.

The IPFS client will cache the requests as needed.

**Comment by RubenKelevra** [comment]
Short update: I decided to give this project a go. Just to avoid that we're working separately on the same thing. More info soon.

**Comment by RubenKelevra** [comment]
Short intermediate status:

I've registered two domains for the cluster for human readability:

> pacman.rocks
> pacman.store

Which do you guys prefer?

Planning of details is done, currently writing the scripts and the guides for usage by endusers as well as a joining guide for people who like to contribute to the cluster.

Should be all done by the end of the week.

**Comment by RubenKelevra** [comment]
Hey guys!

End of the week update (I'm at UTC+01:00): 

There will be a delay - I need another 3-4 hours to complete the work I think. See you tomorrow!

**Comment by RubenKelevra** [comment]
Alright, the first import is running (via script). I'll check tomorrow the delta update part. It should be just minor fixes necessary. Then the cluster just need a configuration file and we can start :)

![Screenshot_20200128_060515](https://user-images.githubusercontent.com/614929/73236987-83c65580-4194-11ea-9021-cb44b89d0fc2.png)

**Comment by RubenKelevra** [comment]
Alright guys. Repo size is around 46 GB.

You need to increase your IPFS cache size accordingly and make sure, you got enough free space on the drive which holds the IPFS cache.

To avoid that everyone is leeching the full repo from my node, please prefill your local cache by running rsync on a official mirror:

`rsync -rtlvH --delete-after --delay-updates --safe-links URL TMP_DIRECTORY`  
While URL can be every mirror from this list:
https://www.archlinux.org/mirrors/status/

make sure to add `/pool/` at the end of the URL to just get all pkgs.

Then add these pkgs to your IPFS:

`ipfs add --progress --recursive --raw-leaves --wrap-with-directory --pin=0 TMP_DIRECTORY`

After this commando has been successfully completed, you can remove the TMP_DIRECTORY.

Now feel free to pin the current version of the repo:
`ipfs pin add --progress /ipfs/QmP5Hudqawz9qcuogYizUh4qTg6wMmVSLVxBTVQeg8S34s`

This will sync every additional files (for example the ISOs) on your node and keep them stored there (for now).

I will test the cluster setup thoroughly in the next days and keep you posted, how to join.

If you just want to use IPFS for receiving updates, this information will follow after the cluster has been created.

**Comment by hsanjuan** [comment]
> ipfs add --progress --recursive --raw-leaves --wrap-with-directory --pin=0 TMP_DIRECTORY

What is the average file size? It might be a good idea to increase the block size to 1M perhaps? Also `--cid-version=1` since, why not..

**Comment by tobixen** [comment]
@RubenKelevra, do you have any plan on how to sync, update and use it?  Pinning a snapshot of the official mirror does not seem that useful as long as all those details are in place.

**Comment by RubenKelevra** [comment]
> > ipfs add --progress --recursive --raw-leaves --wrap-with-directory --pin=0 TMP_DIRECTORY
> 
> What is the average file size? It might be a good idea to increase the block size to 1M perhaps? Also `--cid-version=1` since, why not..

Since my data is currently using version 0, using version 1 isn't going to help.

> @RubenKelevra, do you have any plan on how to sync, update and use it? Pinning a snapshot of the official mirror does not seem that useful unless all those details are in place.

Yes I have. Pinning the current version will just help that I have not send everything when you join the cluster.

**Comment by mohe2015** [comment]
```
ipfs pin add --progress /ipfs/QmP5Hudqawz9qcuogYizUh4qTg6wMmVSLVxBTVQeg8S34s
Fetched/Processed 103854 nodes
```
@RubenKelevra For me this takes really long - do you know how many nodes there should be? Or do you know why it takes so long? This is my second try and I changed some (experimental) settings so that may also be a cause. I'm using the master of go-ipfs so maybe that's also an issue. Or is this to be expected?

**Comment by RubenKelevra** [comment]
@mohe2015 this is to be expected. Currently running of my development machine which has limited bandwidth. I'm right now setting up a production server with more CPU power and 1 GB/s bandwidth. This should fix the issue. :)

**Comment by RubenKelevra** [comment]
Make sure to unpin the snapshot you might have pinned before:
```ipfs pin rm /ipfs/QmP5Hudqawz9qcuogYizUh4qTg6wMmVSLVxBTVQeg8S34s```

There's now a [pull request](https://github.com/ipfs-cluster/collab.ipfscluster.io/pull/4) for the collab cluster website. And I wrote an [Introduction](https://github.com/RubenKelevra/pacman.store#pacmanstore) for the usage.

The segment _Use the pkg cache with Pacman_ is pretty roughly outlined and will be added when it's ready for prime time together with the info about the pacman wrapper - to actually use the cache on machines which has IPFS.

**Comment by guysv** [comment]
@RubenKelevra shouldn't the unpin be recursive?

**Comment by RubenKelevra** [comment]
> @RubenKelevra shouldn't the unpin be recursive?
> Thanks for your initiative btw! glad to be a part of the effort :)

It should be by default, `ipfs pin rm --help` states:

> Removes the pin from the given object allowing it to be garbage
>   collected if needed. (By default, recursively. Use -r=false for direct pins.)

**Comment by RubenKelevra** [comment]
Status-Update:

I reworked the scripts a bit, you can now find a html page with all snapshots of older repo versions [here](https://ipfs.io/ipns/old.pkg.pacman.store/arch/x86_64/default.html).

I also reworked the structure, to allow for easier browsing: 'cache' holds all pkgs, the individual repos have seperate folders and 'db' hold just the db files.

Current version can always be find [here](https://ipfs.io/ipns/pkg.pacman.store/arch/x86_64/default/).

You can join the cluster (you need about 50 GB of storage currently, plus overhead and it will increase in the next 2 month, by all new packages which are released - expect at least 100 GB for this cluster!). Infos can be found [here](https://github.com/RubenKelevra/pacman.store#join-the-cluster).

Three people already joined the cluster - thank you! 

The included FUSE mount option of ipfs has been proven to unstable to rely on (had some weird is mounted, can be accessed but not new stuff and cannot be unmounted etc.). So the temporary "how to use it as client" which also is published in this repo isn't recommended to be used except for testing.

I found a python script which does the same and will be used for this. I also plan to create some aur packages which should allow for easy setup the client stuff on a machine - so no manual configuration is necessary.

**Comment by RubenKelevra** [comment]
Some meta discussions going on in a ticket of my project... if it's a good idea to use the ipfs.io http gateway to download the updates and if it's good to encourage other users to do this:

https://github.com/RubenKelevra/pacman.store/issues/11#issuecomment-589382162

**Comment by RubenKelevra** [comment]
Note to anybody who had already joined the cluster a while back: The cluster setup ran into a non-recoverable state because of a bug and would not replicate anything from the master.

Therefore there's a new secret and a new name for the cluster, restart following the cluster with the command quoted above.

I tried to remove all old pins from the old cluster before shutdown, but this might have not worked either. So check if you're having still local pins in your IPFS-Client.

If you have nothing running in the IPFS-Client except the cluster follower, you can safely remove all pinned entries and start with the command above again.

The following command will unpin everything from your IPFS-Client:
```
ipfs pin ls --type=recursive | awk '{ print $1 }' | xargs -d '\n' ipfs pin rm
```
(may take some minutes)

**Comment by RubenKelevra** [comment]
### If you run multiple machines in a network:

#### MDNS-Method:

Make sure both IPFS-Daemons have mDNS activated. They will connect to each other and exchange the data when needed.

#### Single IPFS-Client as Gateway-Method:

You can change the listening address for the web gateway in the settings of the IPFS-Client which should serve as local package cache. It won't 'cache' in the traditional sense, that you get older outdated content, it will always fetch the latest content directly from the IPFS.

But if there are multiple machines downloading the same updates, they will be already stored locally on the IPFS-Client which runs the Gateway and just have to be transferred in the local network.

Make sure to increase the cache size for this application, since there might be more than 8 GB of different packages installed in your network, so no redownload will be necessary.

If you have a large number of clients and want to prefetch the entire repo to the IPFS-Cache of your local web gateway to speedup updates, even more, you can do this by running this script:

Note that this script has to dig through a lot of data, so there's a substantial amount of IO going on for each run which finds a new update. Make sure to run this on an SSD, not an HDD.

https://github.com/RubenKelevra/pacman.store/blob/master/toolset/follow_cluster_parts.sh

**Comment by dontdieych** [comment]
@RubenKelevra , I've got this error messages.

```
❯ ipfs-cluster-follow pacman.store run --init cluster.pacman.store
cluster.pacman.store will be assumed to be an DNSLink-powered address: /ipns/cluster.pacman.store.
It will be resolved using the local IPFS daemon's gateway (127.0.0.1:8080).
If this is not the case, specify the full url starting with http:// or https://.
(You can override the gateway URL by setting IPFS_GATEWAY)

Identity written to /home/a/.ipfs-cluster-follow/pacman.store/identity.json.
Configuration written to /home/a/.ipfs-cluster-follow/pacman.store/service.json.
Cluster "pacman.store" follower peer initialized.

You can now use "ipfs-cluster-follow pacman.store run" to start a follower peer for this cluster.
(Remember to start your IPFS daemon before)
Starting the IPFS Cluster follower peer for "pacman.store".
CTRL-C to stop it.
Checking if IPFS is online (will wait for 2 minutes)...
waiting for IPFS to become available on /ip4/127.0.0.1/tcp/5001...
06:26:39.553  INFO     config: loading configuration from http://127.0.0.1:8080/ipns/cluster.pacman.store config.go:359
reading the configurations in /home/a/.ipfs-cluster-follow/pacman.store: unsuccessful request (404): 404 page not found


❯ ipfs-cluster-follow pacman.store run
Starting the IPFS Cluster follower peer for "pacman.store".
CTRL-C to stop it.
Checking if IPFS is online (will wait for 2 minutes)...
waiting for IPFS to become available on /ip4/127.0.0.1/tcp/5001...
06:28:17.897  INFO     config: loading configuration from http://127.0.0.1:8080/ipns/cluster.pacman.store config.go:359
reading the configurations in /home/a/.ipfs-cluster-follow/pacman.store: unsuccessful request (404): 404 page not found

```

**Comment by dontdieych** [comment]
> @RubenKelevra , I've got this error messages.
> 
> ```
> ❯ ipfs-cluster-follow pacman.store run --init cluster.pacman.store
> cluster.pacman.store will be assumed to be an DNSLink-powered address: /ipns/cluster.pacman.store.
> It will be resolved using the local IPFS daemon's gateway (127.0.0.1:8080).
> If this is not the case, specify the full url starting with http:// or https://.
> (You can override the gateway URL by setting IPFS_GATEWAY)
> 
> Identity written to /home/a/.ipfs-cluster-follow/pacman.store/identity.json.
> Configuration written to /home/a/.ipfs-cluster-follow/pacman.store/service.json.
> Cluster "pacman.store" follower peer initialized.
> 
> You can now use "ipfs-cluster-follow pacman.store run" to start a follower peer for this cluster.
> (Remember to start your IPFS daemon before)
> Starting the IPFS Cluster follower peer for "pacman.store".
> CTRL-C to stop it.
> Checking if IPFS is online (will wait for 2 minutes)...
> waiting for IPFS to become available on /ip4/127.0.0.1/tcp/5001...
> 06:26:39.553  INFO     config: loading configuration from http://127.0.0.1:8080/ipns/cluster.pacman.store config.go:359
> reading the configurations in /home/a/.ipfs-cluster-follow/pacman.store: unsuccessful request (404): 404 page not found
> 
> 
> ❯ ipfs-cluster-follow pacman.store run
> Starting the IPFS Cluster follower peer for "pacman.store".
> CTRL-C to stop it.
> Checking if IPFS is online (will wait for 2 minutes)...
> waiting for IPFS to become available on /ip4/127.0.0.1/tcp/5001...
> 06:28:17.897  INFO     config: loading configuration from http://127.0.0.1:8080/ipns/cluster.pacman.store config.go:359
> reading the configurations in /home/a/.ipfs-cluster-follow/pacman.store: unsuccessful request (404): 404 page not found
> ```

Errors gone after changing default port on config file. I use syncthing daemon(8080) so changed ipfs gateway port to 8082.

**Comment by RubenKelevra** [comment]
> > @RubenKelevra , I've got this error messages.
> > ```
> > ❯ ipfs-cluster-follow pacman.store run --init cluster.pacman.store
> > cluster.pacman.store will be assumed to be an DNSLink-powered address: /ipns/cluster.pacman.store.
> > It will be resolved using the local IPFS daemon's gateway (127.0.0.1:8080).
> > If this is not the case, specify the full url starting with http:// or https://.
> > (You can override the gateway URL by setting IPFS_GATEWAY)
> > 
> > Identity written to /home/a/.ipfs-cluster-follow/pacman.store/identity.json.
> > Configuration written to /home/a/.ipfs-cluster-follow/pacman.store/service.json.
> > Cluster "pacman.store" follower peer initialized.
> > 
> > You can now use "ipfs-cluster-follow pacman.store run" to start a follower peer for this cluster.
> > (Remember to start your IPFS daemon before)
> > Starting the IPFS Cluster follower peer for "pacman.store".
> > CTRL-C to stop it.
> > Checking if IPFS is online (will wait for 2 minutes)...
> > waiting for IPFS to become available on /ip4/127.0.0.1/tcp/5001...
> > 06:26:39.553  INFO     config: loading configuration from http://127.0.0.1:8080/ipns/cluster.pacman.store config.go:359
> > reading the configurations in /home/a/.ipfs-cluster-follow/pacman.store: unsuccessful request (404): 404 page not found
> > 
> > 
> > ❯ ipfs-cluster-follow pacman.store run
> > Starting the IPFS Cluster follower peer for "pacman.store".
> > CTRL-C to stop it.
> > Checking if IPFS is online (will wait for 2 minutes)...
> > waiting for IPFS to become available on /ip4/127.0.0.1/tcp/5001...
> > 06:28:17.897  INFO     config: loading configuration from http://127.0.0.1:8080/ipns/cluster.pacman.store config.go:359
> > reading the configurations in /home/a/.ipfs-cluster-follow/pacman.store: unsuccessful request (404): 404 page not found
> > ```
> 
> Errors gone after changing default port on config file. I use syncthing daemon(8080) so changed ipfs gateway port to 8082.

I'll take a look at my setup, to confirm everything is working on my end, will report back in some minutes, thanks for the report!

If you use a different port, you have to set an environment variable, before you launch ipfs-cluster-follower:

```
The IPFS_GATEWAY environmental variable can be used to set the gateway location if it’s not the default (127.0.0.1:8080).
```

More info on that can be found here:

https://cluster.ipfs.io/documentation/reference/follow/

Just a sidenote: The cluster holds a large amount of informations, so the initial "catch up" to the current state, will take some time and cannot be interrupted. If you need to interrupt early, it's best to start from the beginning again.

Once this is finished, you can interrupt at any time.

This will be improved in the near future when ipfs-cluster has implemented commits, which will pack multiple changes into one 'change' of the cluster state.

**Comment by RubenKelevra** [comment]
@dontdieych So I've tested my setup. When you change the port to the local IPFS-Instance it _should_ work flawlessly. If you still run into troubles, just send me an email. :)

**Comment by dontdieych** [comment]
@RubenKelevra Thanks. It looks like working.

Another question. Is there a way to download pkgs automatically that already installed? Oviously when new version appeared.

**Comment by RubenKelevra** [comment]
@dontdieych yes, you can simply run a `pacman -Syuw --noconfirm` as a cronjob.

This will issue a download request for everything new, and download them to your local cache.

**Comment by dontdieych** [comment]
@RubenKelevra Thanks! again. :+1:

**Comment by RubenKelevra** [comment]
I've started a discussion if it wouldn't make sense to extend pacman natively with the ability to request packages from a running IPFS daemon via the API.

To be able to do this there's a need to extend the database-format with the Content-ID of each package (currently they hold a MD5 and a SHA256 sum plus a signature).

This way there would be no noticeable download delay since there's no need to fetch and open large folders (which is currently the main reason why my approach is slow).

The discussion is on the pacman-development mailing list, if somebody wants to follow:

https://lists.archlinux.org/pipermail/pacman-dev/2020-April/024179.html

**Comment by RubenKelevra** [comment]
> I've started a discussion if it wouldn't make sense to extend pacman natively with the ability to request packages from a running IPFS daemon via the API.
> 
> To be able to do this there's a need to extend the database-format with the Content-ID of each package (currently they hold an MD5 and a SHA256 sum plus a signature).
> 
> This way there would be no noticeable download delay since there's no need to fetch and open large folders (which is currently the main reason why my approach is slow).
> 
> The discussion is on the pacman-development mailing list if somebody wants to follow:
> 
> https://lists.archlinux.org/pipermail/pacman-dev/2020-April/024179.html

This discussion has ceased, but if you like to add notes or track the status long term, you can find a ticket here: https://github.com/RubenKelevra/pacman.store/issues/41

**Comment by RubenKelevra** [comment]
Hey @jbenet I think we can close this one, right? :)

**Comment by RubenKelevra** [comment]
Off-Topic: What do you guys think about the idea of asking the Flatpak project to integrate IPFS?

https://github.com/ipfs/notes/issues/444

**Comment by RubenKelevra** [comment]
I still don't have another method of reaching you all, so I add this here:

Make sure to update your cluster nodes to 0.14.0 :)

**Comment by RubenKelevra** [comment]
In case someone likes Endeavouros - the small additional repository of this project is now available on the pacman.store collab cluster as well:

> I've added it to the cluster. It can be viewed in the Browser via
> 
> http://endeavouros.pkg.pacman.store/
> 
> For pacman you just have to configure it to use the local ipfs node, which should look like this:
> 
> `Server = http://endeavouros.pkg.pacman.store.ipns.localhost:8080/repo/endeavouros/x86_64`

_Originally posted by @RubenKelevra in https://github.com/endeavouros-team/mirrors/issues/1#issuecomment-895964369_

**Comment by OpenOS-Project-Creator-Founder** [comment]
https://github.com/OpenOS-Project-Creator-Founder/ipos

https://github.com/OpenOS-Project-Creator-Founder/astralboot

https://github.com/OpenOS-Project-Creator-Founder/netboot.ipfs

**Comment by OpenOS-Project-Creator-Founder** [comment]
https://github.com/ipfs/notes/issues/84#issuecomment-1347453802
